### PR TITLE
Cleanup metrics-node-sampler code

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

File: pkg/sampler/server.go [hunk 1]
```diff
@@ -218,60 +244,251 @@ func (s *Server) Start(ctx context.Context, stop context.CancelFunc) error {
...
+	var err error
+	var badConnections []string
+
+	// Send metrics to each collector
+	for _, c := range collectors {
+		start := time.Now()
+
+		// clone the response because we send a different one to each collector
+		retries := 0
+		for ; retries < s.SendPushMetricsRetryCount+1; retries++ {
```
Comment (campoy) [inline pkg/sampler/server.go:428]:
```suggestion
		for ; retries <= s.SendPushMetricsRetryCount; retries++ {
```

File: pkg/sampler/server.go [hunk 2]
```diff
@@ -149,7 +171,11 @@ func (s *Server) Start(ctx context.Context, stop context.CancelFunc) error {
 	go func() {
 		errs <- errors.WithStack(s.cache.Start(ctx))
 	}()
+	return nil
+}
 
+// startGRPC starts the GRPC server for receiving registration and list requests
+func (s *Server) startGRPC(ctx context.Context, errs chan error) error {
```
Comment (campoy) [inline pkg/sampler/server.go:178]:
```suggestion
func (s *Server) startGRPC(ctx context.Context, errs chan<- error) error {
```

**Review by campoy — APPROVED** [review]
A bunch of nits and questions, but no objection to the change.

File: pkg/sampler/server.go [hunk 3]
```diff
@@ -218,60 +244,251 @@ func (s *Server) Start(ctx context.Context, stop context.CancelFunc) error {
...
-		// find the list of servers to push metrics to
-		ips, err := net.LookupIP(s.PushHeadlessService)
-		if err != nil {
-			if strings.Contains(err.Error(), "no such host") {
-				log.Info("unable to lookup collector servers from dns records",
-					"service", s.PushHeadlessService, "msg", err.Error())
-			} else {
-				log.Error(err, "unable to lookup collector servers from dns records", "service", s.PushHeadlessService)
+func (s *Server) startRegisterCollectorsFromDNS() {
+	go func() {
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply pkg/sampler/server.go:396]: done

File: pkg/sampler/server.go [hunk 4]
```diff
@@ -129,7 +129,29 @@ func (s *Server) Start(ctx context.Context, stop context.CancelFunc) error {
...
+		log.Error(err, "stopping node-sampler due to error")
+		return err
+	case <-ctx.Done():
+		log.Info("stopping node-sampler due to context done", "err", ctx.Err())
+		return nil
+	}
+}
+
+// startCache starts the cache for caching utilization samples
+func (s *Server) startCache(ctx context.Context, errs chan error) error {
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply pkg/sampler/server.go:152]: done

File: pkg/sampler/server.go [hunk 5]
```diff
@@ -218,60 +244,251 @@ func (s *Server) Start(ctx context.Context, stop context.CancelFunc) error {
...
+				"addr", addr,
+				"reason", sendCtx.Err(),
+				"second", time.Since(start).Seconds())
+			s.removeConnections([]string{addr})
+		}()
+	}
+}
+
+// removeConnections stops existing connections
+func (s *Server) removeConnections(ips []string) {
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply pkg/sampler/server.go:359]: done

File: pkg/sampler/server.go [hunk 6]
```diff
@@ -298,127 +515,47 @@ func (s *Server) foundUnsentPods(lastResp *api.ListMetricsResponse) bool {
...
+
+		ticker := time.NewTicker(s.pushFrequency)
+		createdPodTicker := time.NewTicker(s.checkCreatedPodFrequency)
+		defer func() {
+			ticker.Stop()
+			createdPodTicker.Stop()
+			s.shutdownAllConnections()
+		}() // send EOF to all collectors
+
+		// try the "first" request for 1 minute
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply pkg/sampler/server.go:531]: Updated.  Just dropping that for now to simplify the code.

File: pkg/sampler/server.go (hunk 1 above)
Comment (pwittrock) [reply pkg/sampler/server.go:428]: done

File: pkg/sampler/server.go (hunk 2 above)
Comment (pwittrock) [reply pkg/sampler/server.go:178]: done

**Pull request merged**